README.md

```diff
@@ -58,81 +58,101 @@ Include this badge in your readme if you make a new module that implements inter
 
-#### `util.serialize(dagNode, callback)`
 
-> serializes a dagNode of an IPLD format into its binary format
+### serialize(IpldNode)
 
-`callback` must have the signature `function (err, binaryBlob)`, where `err` is an Error is the function fails and `binaryBlob` is a Buffer containing the serialized version.
+> Serializes the internal representation of an IPLD Node into a binary blob.
 
-#### `util.deserialize(binaryBlob, callback)`
+`IpldNode` is a previously deserialized binary Block.
 
-> deserializes a binary blob into the instance
+Returns a Promise containing a `Buffer` with the serialized version of the given IPLD Node.
 
-`callback` must have the signature `function (err, dagNode)`, where `err` is an Error if the function fails and `dagNode` is the dagNode that got deserialized in the process.
 
-#### `util.cid(binaryBlob[, options], callback)`
+### deserialize(binaryBlob)
 
-> get the CID of a binary blob
+> Deserialize into internal representation.
 
-Options include:
-  - version - the CID version to be used (defaults to 1)
-  - hashAlg - the hash algorithm to be used (default to the one set by the format)
+The result is a JavaScript object, which may also be a Proxy object in case the data shouldn’t be deserialized as a whole. Its fields are the public API that can be resolved through. It’s up to the format to add convenient methods for manipulating the data. The returned object may also be a Proxy object in case the data shouldn’t be deserialized as a whole.
 
-`callback` must have the signature `function (err, cid)`, where `err` is an Error if the function fails and `cid` is a CID instance of the binary blob.
+Returns a Promise containing the Javascript object. This object must be able to be serialized with a `serialize()` call.
 
-### Local resolver methods
 
-#### `resolver.resolve(binaryBlob, path, callback)`
+### cid(binaryBlob, [options])
 
-> resolves a path in block, returns the value and or a link and the partial missing path. This way the IPLD Resolver can fetch the link and continue to resolve.
+> Return the CID of the binary blob.
 
-`callback` must have the signature `function (err, result)`, where `err` is an Error if the function fails and `result` is an object with the following keys:
+Possible `options` are:
+  - `version` (`number`, default: 1): the CID version to be used
+  - `hashAlg` (`Multicodec`, default: the one the format specifies): the hash algorithm to be used
 
-- value: <> - The value resolved or an IPLD link if it was unable to resolve it through.
-- remainderPath: <> - The remaining path that was not resolved under block scope.
+This can be used to verify that some data actually has a certain CID.
 
-If `path` is the root `/`, the result is a nested object that contains all paths that `tree()` returns. The values are the same as accessing them directly with the full path. Example:
 
-`tree()` returns:
+### toJSON(IpldNode)
```

Comment:
So, we cut the "canonical serialization" stuff from IPLD a while back. That means this method probably shouldn't be required and that any conversions from one format to another are custom.

Given that we don't have a method that takes this representation and turns it back into a serialized block should we include this?

Reply:
The thing is, we've sort of opened the door to formats having types that they can't easily, or potentially don't want to, convert to a JSON representation.

Even the base types we require for the data model are *base* requirements, there's nothing keeping a format from have more types and there's no guarantee you could represent those in JSON.

Rather than put this on the codec's to write I would just write something for "search on IPLD" that converts an object to a JSON representation and is aware of Binary and CID types. That will work with anything that supports the data model and you could just pass it the deserialized node.

Reply:
Very good points. I was just so keen to have a JSON representation. But it can indeed live outside of IPLD Format API which should really be kept to a minimum. I'll remove that part.